### PR TITLE
Editorial suggestions from Benjamin Kaduk (transport)

### PR DESCRIPTION
**Pull request by kaduk**
I made some proposed tweaks in my local copy as I was doing my IESG review.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1684,7 +1686,9 @@ retry_source_connection_id transport parameter.
 
 The values provided by a peer for these transport parameters MUST match the
 values that an endpoint used in the Destination and Source Connection ID fields
-of Initial packets that it sent. Including connection ID values in transport
+of Initial packets that it sent (and received, for servers). Endpoints MUST
+validate that received transport parameters match received Connection ID values.
+Including connection ID values in transport
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:1691]: I'm calling this "editorial" even though it adds a new MUST, because I believe the intent was already that you do this, and so the text is just reiterating an existing expectation.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -4739,8 +4747,9 @@ server MAY send multiple Initial packets.  The cryptographic key exchange could
 require multiple round trips or retransmissions of this data.
 
 The payload of an Initial packet includes a CRYPTO frame (or frames) containing
-a cryptographic handshake message, ACK frames, or both.  PING, PADDING, and
-CONNECTION_CLOSE frames of type 0x1c are also permitted.  An endpoint that
+a cryptographic handshake message, ACK frames, CONNECTION_CLOSE frames of type
+0x1c, or some combination thereof.  PING and PADDING frames are also permitted.
+An endpoint that
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:4752]: This is pretty awkward.  The idea is that the old text implies that it is invalid to have an Initial with neither a CRYPTO nor an ACK.  But IIUC it's valid to send CONNECTION_CLOSE and nothing else...

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -5065,8 +5075,8 @@ Destination Connection ID:
 
 Packet Number:
 
-: The packet number field is 1 to 4 bytes long. The packet number has
-  confidentiality protection separate from packet protection, as described in
+: The packet number field is 1 to 4 bytes long. The packet number is protected
+  using header protection; see
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:5079]: Trying to use consistent language for each field covered by header protection.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -5257,7 +5267,8 @@ initial_max_stream_data_uni (0x07):
 initial_max_streams_bidi (0x08):
 
 : The initial maximum bidirectional streams parameter is an integer value that
-  contains the initial maximum number of bidirectional streams the peer is
+  contains the initial maximum number of bidirectional streams the endpoint
+  that receives this transport parameter is
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:5271]: We were inconsistent about how we described the two endpoints in these paragraphs.  I'm not sure I caught all instances, though, but the "the endpoint that receives this transport parameter" seems like a good clear description.

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -5437,7 +5449,7 @@ application protocol wishes to prevent the connection from timing out; see
 Receivers send ACK frames (types 0x02 and 0x03) to inform senders of packets
 they have received and processed. The ACK frame contains one or more ACK Ranges.
 ACK Ranges identify acknowledged packets. If the frame type is 0x03, ACK frames
-also contain the sum of QUIC packets with associated ECN marks received on the
+also contain the cumulative count of QUIC packets with associated ECN marks received on the
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:5452]: (you can't take a sum of QUIC packets; there is no addition operation defined for them.)

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -5492,15 +5504,14 @@ ACK Delay:
 
 ACK Range Count:
 
-: A variable-length integer specifying the number of Gap and ACK Range fields in
+: A variable-length integer specifying the number of ACK Range fields in
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:5507]: This whole section was a bit awkward.  I tried to operate on the principle that the Ack Range is the compound structure, and that Gap and the Ack Range Length are the member fields within it.  The resulting text isn't very pretty, though...

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -6520,8 +6531,8 @@ of the security properties of QUIC depend on the TLS handshake providing these
 properties. Any attack on the TLS handshake could affect QUIC.
 
 Any attack on the TLS handshake that compromises the secrecy or uniqueness
-of session keys affects other security guarantees provided by QUIC that depends
-on these keys. For instance, migration ({{migration}}) depends on the efficacy
+of session keys, or the authentication of the participating peers, affects other security guarantees provided by QUIC that depend
+on those keys. For instance, migration ({{migration}}) depends on the efficacy
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:6535]: I'm open to better phrasing suggestions here, too.

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -6863,13 +6877,16 @@ Servers SHOULD provide mitigations for this attack by limiting the usage and
...
-## Optimistic ACK Attack
+## Optimistic ACK Attack (#optimistic-ack-attack}
 
 An endpoint that acknowledges packets it has not received might cause a
 congestion controller to permit sending at rates beyond what the network
 supports.  An endpoint MAY skip packet numbers when sending packets to detect
 this behavior.  An endpoint can then immediately close the connection with a
 connection error of type PROTOCOL_VIOLATION; see {{immediate-close}}.
+Note that introducing artificial gaps might increase the size of the
+returned ACK packets in a manner that scales roughly llinearly with the
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:6888]: Oops, "llinearly" typo.
```suggestion
returned ACK packets in a manner that scales roughly linearly with the
```

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -7540,7 +7559,7 @@ Description:
 
 The initial contents of this registry are shown in {{iana-error-table}}.
 
-| Value | Error                     | Description                   | Specification   |
+| Value | Code                      | Description                   | Specification   |
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:7562]: (The prose says this field is 'Code', not 'Error'.)

File: draft-ietf-quic-transport.md (hunk 7 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md:6535]: The structure doesn't seem ideal, but this is OK with me and I can't see anything better.

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -6585,22 +6596,25 @@ encrypted with keys derived from the TLS key exchange.  Further, parameter
 negotiation is folded into the TLS transcript and thus provides the same
 integrity guarantees as ordinary TLS negotiation.  An attacker can observe
 the client's transport parameters (as long as it knows the version-specific
-salt) but cannot observe the server's transport parameters and cannot influence
-parameter negotiation.
+salt) but cannot observe the server's transport parameters as sent to
+the client and cannot influence
+parameter negotiation.  (The attacker could make its own connection to
+the server and observe the transport parameters sent on that
+connection.)
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:6603]: I don't think that this adds much value and so I would prefer not to take this change.

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -6585,22 +6596,25 @@ encrypted with keys derived from the TLS key exchange.  Further, parameter
...
 
 Connection IDs are unencrypted but integrity protected in all packets.
 
-This version of QUIC does not incorporate a version negotiation mechanism;
+This version of QUIC does not incorporate a complete or secure version negotiation mechanism;
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:6607]: This is unnecessary.

File: draft-ietf-quic-transport.md [hunk 12]
```diff
@@ -6613,7 +6627,7 @@ ensures that valid packets are only generated by endpoints that possess the
 key material established during the handshake; see {{handshake}} and
 {{handshake-properties}}. Similarly, any active attacker that observes packets
 and attempts to insert new data or modify existing data in those packets should
-not be able to generate packets deemed valid by the receiving endpoint.
+not be able to generate non-Initial packets deemed valid by the receiving endpoint.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:6630]:
```suggestion
not be able to generate packets deemed valid by the receiving endpoint,
other than Initial packets.
```

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -6658,7 +6672,7 @@ An on-path attacker can:
 - Delay packets
 - Reorder packets
 - Drop packets
-- Split and merge datagrams along packet boundaries
+- Split and merge datagrams along packet boundaries (though header protection obfsucates the location of packet boundaries)
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:6675]: Unnecessary change.

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -6863,13 +6877,16 @@ Servers SHOULD provide mitigations for this attack by limiting the usage and
...
+## Optimistic ACK Attack (#optimistic-ack-attack}
 
 An endpoint that acknowledges packets it has not received might cause a
 congestion controller to permit sending at rates beyond what the network
 supports.  An endpoint MAY skip packet numbers when sending packets to detect
 this behavior.  An endpoint can then immediately close the connection with a
 connection error of type PROTOCOL_VIOLATION; see {{immediate-close}}.
+Note that introducing artificial gaps might increase the size of the
+returned ACK packets in a manner that scales roughly llinearly with the
+number of gaps introduced, so fewer, larger, gaps may be more efficient.
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md:6889]: Unnecessary change.

**Review by martinthomson — CHANGES_REQUESTED** [review]
I've some concerns with some of the changes here, but there are a lot of good ones.  Some of these are unnecessary however.  Can you revert or fix those changes I've identified?  Thanks.

File: draft-ietf-quic-transport.md (hunk 2 above)
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Done.

File: draft-ietf-quic-transport.md [hunk 15]
```diff
@@ -4895,7 +4904,8 @@ Retry Packet {
 
 A Retry packet (shown in {{retry-format}}) does not contain any protected
 fields.  The value in the Unused field is set to an arbitrary value by the
-server; a client MUST ignore these bits.  In addition to the fields from the
+server; a client MUST NOT process those bits other than to verify the Retry
+Integrity Tag.  In addition to the fields from the
```
[the remark this answers is not in the document]
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Okay.  I think we have ample precedent for "process" being a flexible term...

File: draft-ietf-quic-transport.md (hunk 10 above)
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Okay; reverted.

File: draft-ietf-quic-transport.md (hunk 11 above)
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Okay; reverted.

File: draft-ietf-quic-transport.md (hunk 13 above)
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Reverted

File: draft-ietf-quic-transport.md (hunk 14 above)
Comment (kaduk) [reply draft-ietf-quic-transport.md]: Reverted.

**Review by kaduk — COMMENTED** [review]
I think I made all the requested changes.  I will make a pass through the diff to check that I didn't introduce any weirdness as a result.

File: draft-ietf-quic-transport.md [hunk 16]
```diff
@@ -6863,7 +6872,7 @@ Servers SHOULD provide mitigations for this attack by limiting the usage and
 lifetime of address validation tokens; see {{validate-future}}.
 
 
-## Optimistic ACK Attack
+## Optimistic ACK Attack (#optimistic-ack-attack}
```
Comment (kaduk) [inline draft-ietf-quic-transport.md:6875]:
```suggestion
## Optimistic ACK Attack {#optimistic-ack-attack}
```

**Comment by kaduk** [comment]
I would like some help with the CI issues (I don't currently have a CircleCI account and would prefer to not have to read the T&C just for this pull request).

**Comment by LPardue** [comment]
@kaduk addiction the failure is simply line length. If you hard wrap everything to 80 characters all should be good.

**Comment by kaduk** [comment]
@LPardue ah, thanks.

**Review by martinthomson — APPROVED** [review]

File: draft-ietf-quic-transport.md [hunk 17]
```diff
@@ -5293,8 +5303,8 @@ disable_active_migration (0x0c):
 
 : The disable active migration transport parameter is included if the endpoint
   does not support active connection migration ({{migration}}) on the address
-  being used during the handshake.  When a peer sets this transport parameter,
-  an endpoint MUST NOT use a new local address when sending to the address that
+  being used during the handshake.  An endpoint that receives this transport
+  parameter MUST NOT use a a new local address when sending to the address that
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:5307]:
```suggestion
  parameter MUST NOT use a new local address when sending to the address that
```

File: draft-ietf-quic-transport.md (hunk 6 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md:5506]: This is a good improvement -- thanks!

**Review by janaiyengar — APPROVED** [review]
A couple of nits, which I'll commit, but otherwise lgtm.
